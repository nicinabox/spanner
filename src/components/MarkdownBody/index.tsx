export * from './MarkdownBody';
export { default } from './MarkdownBody';
