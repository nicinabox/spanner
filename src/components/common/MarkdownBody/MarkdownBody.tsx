import { Box } from '@chakra-ui/react';
import { marked } from 'marked';
import Head from 'next/head';
import React from 'react';

export interface MarkdownBodyProps {
    body: string;
}

const styles = {
    '.chakra-ui-dark &': {
        color: 'white',
    },
    '.chakra-ui-dark & tr': {
        backgroundColor: 'gray.800',
    },
    '.chakra-ui-dark & tr:nth-of-type(2n)': {
        backgroundColor: 'gray.700',
    },
};

export const MarkdownBody: React.FC<MarkdownBodyProps> = ({ body }) => (
    <>
        <Head>
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/github-markdown-css/4.0.0/github-markdown.min.css" integrity="sha512-Oy18vBnbSJkXTndr2n6lDMO5NN31UljR8e/ICzVPrGpSud4Gkckb8yUpqhKuUNoE+o9gAb4O/rAxxw1ojyUVzg==" crossOrigin="anonymous" />
        </Head>
        <Box className="markdown-body" sx={styles} fontFamily="inherit" dangerouslySetInnerHTML={{ __html: marked(body) }} />
    </>
);

export default MarkdownBody;
