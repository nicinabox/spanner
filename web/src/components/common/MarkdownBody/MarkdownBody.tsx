import { Box } from '@chakra-ui/react';
import { marked } from 'marked';
import Head from 'next/head';
import React from 'react';

export interface MarkdownBodyProps {
    body: string;
}

// @ts-expect-error multiple renders
const renderer: marked.Renderer = {
    table(header, body) {
        return `
        <div style="overflow-x: auto;">
            <table style="width: 100%;">
                ${header}
                ${body}
            </table>
        </div>
        `;
    },
};

marked.use({ renderer });

const styles = {
    '&': {
        backgroundColor: 'transparent',
    },
    '& a': {
        color: 'brand.primary',
        textDecoration: 'underline',
    },
    '& table tr': {
        backgroundColor: 'transparent',
    },
    '& table tr:nth-of-type(2n), & thead': {
        backgroundColor: 'brand.100',
    },
    '.chakra-ui-dark &': {
        color: 'white',
    },
    '.chakra-ui-dark & a': {
        color: 'brandInverted.primary',
        textDecoration: 'underline',
    },
    '.chakra-ui-dark & tr:nth-of-type(2n), .chakra-ui-dark & thead': {
        backgroundColor: 'gray.800',
    },
};

export const MarkdownBody: React.FC<MarkdownBodyProps> = ({ body }) => {
    return (
        <>
            <Head>
                <link
                    rel="stylesheet"
                    href="https://cdnjs.cloudflare.com/ajax/libs/github-markdown-css/5.8.1/github-markdown.min.css"
                    integrity="sha512-BrOPA520KmDMqieeM7XFe6a3u3Sb3F1JBaQnrIAmWg3EYrciJ+Qqe6ZcKCdfPv26rGcgTrJnZ/IdQEct8h3Zhw=="
                    crossOrigin="anonymous"
                />
            </Head>
            <Box
                className="markdown-body"
                sx={styles}
                fontFamily="inherit"
                dangerouslySetInnerHTML={{ __html: marked(body) }}
            />
        </>
    );
};

export default MarkdownBody;
