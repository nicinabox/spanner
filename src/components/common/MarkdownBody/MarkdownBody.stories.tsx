import React from 'react';

import { Meta } from '@storybook/react';

import { MarkdownBody, MarkdownBodyProps } from './MarkdownBody';

const Template: React.VFC<MarkdownBodyProps> = (props) => <MarkdownBody {...props} />;

export const Default = Template.bind({});
Default.args = {
    body: '# Hello world',
};

export default {
    title: 'Components/MarkdownBody',
    component: MarkdownBody,
} as Meta;
