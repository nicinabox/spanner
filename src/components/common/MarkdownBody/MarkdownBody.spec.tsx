import React from 'react';

import { render } from '@testing-library/react';

import MarkdownBody from '.';

describe('MarkdownBody', () => {
    it('renders markdown to html', () => {
        const props = {
            body: '# Hello World',
        };
        const { container } = render(<MarkdownBody {...props} />);
        expect(container).toMatchSnapshot();
    });
});
