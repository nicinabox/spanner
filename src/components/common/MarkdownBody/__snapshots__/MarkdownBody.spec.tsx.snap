// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownBody renders markdown to html 1`] = `
<div>
  <div
    class="markdown-body css-1qb95kv"
  >
    <h1
      id="hello-world"
    >
      Hello World
    </h1>
    

  </div>
</div>
`;
